car_control.py

```diff
@@ -171,31 +162,89 @@ def update(self, line):
             self.throttle_old_result = throttle_new_result
     
             throttle_p_output = throttle_new_result # Standard PID Stuff here... nothing particularly interesting :)
-            self.throttle_i_output = max(min(self.throttle_i_output + throttle_new_result, THROTTLE_I_MAX), THROTTLE_I_MIN)
+            self.throttle_i_output = max(min(self.throttle_i_output + throttle_new_result, self.THROTTLE_I_MAX), self.THROTTLE_I_MIN)
             throttle_d_output = ((throttle_delta_result * 1000) / self.fps.delta()) if self.fps.delta() else 0
-            throttle_pid_output = (THROTTLE_P_GAIN * throttle_p_output) + \
-                                  (THROTTLE_I_GAIN * self.throttle_i_output) + \
-                                  (THROTTLE_D_GAIN * throttle_d_output)
+            throttle_pid_output = (self.THROTTLE_P_GAIN * throttle_p_output) + \
+                                  (self.THROTTLE_I_GAIN * self.throttle_i_output) + \
+                                  (self.THROTTLE_D_GAIN * throttle_d_output)
     
             # Throttle goes from 0% to 100%.
             self.throttle_output = max(min(throttle_pid_output, 100), 0)
         else:
             self.throttle_output = self.throttle_output * 0.99
-    
+            self.steering_output = self.STEERING_OFFSET
+            
+        self.jsonreadout()
+        
+        if not self.paused.state:
+            try:
+                client, addr = self.listener.accept()
+            except socket.timeout:
+                pass
+            except:
+                raise
+            else:
+                data = client.recv(1024).decode()
+                if data == 'PAUSE':
+                    print('PAUSE triggered by client', file=self.log_file, flush=True)
+                    self.paused.state = not self.paused.state
+                    client.send(data.encode())
+        
         if self.paused.state:
             self.throttle_output = 0
-            self.steering_output = STEERING_OFFSET
-            time.sleep(0.001)
+            self.steering_output = self.STEERING_OFFSET
+            
+            try:
+                client, addr = self.listener.accept()
+            except socket.timeout:
+                pass
+            except:
+                raise
+            else:
+                data = client.recv(1024).decode()
+                if data == 'UNPAUSE':
+                    print('UNPAUSE triggered by client', file=self.log_file, flush=True)
+                    self.paused.state = not self.paused.state
+                    client.send(data.encode())
+                elif data == 'LOAD':
+                    print('LOAD triggered by client', file=self.log_file, flush=True)
+                    self.load('testjson.json')
```

Comment:
Should this be an if..else instead of two if statements? How long will it wait for a message?

Reply:
The time out in the __init__ method in self.listener.settimeout(0.001) . We will look into changing this into an if.else .

The 0.001 came from the previous sleep command